services/121-service/src/payments/fsp-integration/safaricom/safaricom.api.service.ts

```diff
@@ -1,21 +1,68 @@
+import { TransferRequestSafaricomApiDto } from '@121-service/src/payments/fsp-integration/safaricom/dtos/safaricom-api/transfer-request-safaricom-api.dto';
+import { TransferResponseSafaricomApiDto } from '@121-service/src/payments/fsp-integration/safaricom/dtos/safaricom-api/transfer-response-safaricom-api.dto';
+import { DoTransferReturnType } from '@121-service/src/payments/fsp-integration/safaricom/interfaces/do-transfer-return-type.interface';
+import { SafaricomApiError } from '@121-service/src/payments/fsp-integration/safaricom/safaricom-api.error';
 import { CustomHttpService } from '@121-service/src/shared/services/custom-http.service';
 
+const callbackBaseUrl = process.env.EXTERNAL_121_SERVICE_URL + 'api/';
+const safaricomTimeoutCallbackUrl = `${callbackBaseUrl}financial-service-providers/safaricom/timeout-callback`;
+const safaricomTransferCallbacktUrl = `${callbackBaseUrl}financial-service-providers/safaricom/transfer-callback`;
+
 @Injectable()
 export class SafaricomApiService {
-  public tokenSet: TokenSet | null;
+  public tokenSet: TokenSet;
 
   public constructor(private readonly httpService: CustomHttpService) {}
 
-  public async authenticate(): Promise<string | undefined> {
+  public async transfer({
+    transferAmount,
+    phoneNumber,
+    idNumber,
+    originatorConversationId,
+  }): Promise<DoTransferReturnType> {
+    const payload = this.createTransferPayload({
+      transferAmount,
+      phoneNumber,
+      idNumber,
+      originatorConversationId,
+    });
+    const transferResponse = await this.makeTransferCall(payload);
+
+    let errorMessage: string | undefined;
+
+    if (!transferResponse || !transferResponse.data) {
+      errorMessage = `Error: No response data from Safaricom API`;
+    } else if (transferResponse.data.errorCode) {
+      if (transferResponse.data.errorCode === '500.002.1001') {
+        // This happens only in case of unintended Redis job re-attempt, and only if the API-request already went through the first time
+        // Only console.error and then return to close the job without any further processing
+        console.error(
+          `Error ${transferResponse.data.errorMessage} for originatorConversationId ${originatorConversationId}`,
+        );
+        return { duplicateOriginatorConversationIdError: true };
+      }
+      errorMessage = `${transferResponse.data.errorCode} - ${transferResponse.data.errorMessage}`;
+    } else if (!transferResponse.data.ResponseCode) {
+      errorMessage = `Error: ${transferResponse.data?.statusCode} ${transferResponse.data?.error}`;
+    } else if (transferResponse.data.ResponseCode !== '0') {
+      errorMessage = `Response: ${transferResponse.data?.ResponseCode} - ${transferResponse.data?.ResponseDescription}`;
+    }
+
+    if (errorMessage) {
+      throw new SafaricomApiError(errorMessage);
+    }
+
+    // All the checks above mean that at this stage transferResponse.data.ResponseCode === '0'
```

Comment:
It feels the if structure builds up more intuitively now. Because to check for that at the beginning you need to also check if transferResponse exists and transferResponse.data exists and transferResponse.data.ResponseCode exists, which makes for either longer (writing this all out) or less clean (use questionmarks) code.